lxd/api.go

```diff
@@ -208,12 +214,98 @@ func restServer(d *Daemon) *http.Server {
 		_ = response.NotFound(nil).Render(w)
 	})
 
+	// Initialize API metrics and use middleware functions to update API request metrics.
+	metrics.InitAPIMetrics()
+	mux.Use(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Set the callback function to track the request as completed.
+			// Assure it can be called at most once.
+			var once sync.Once
+			callbackFunc := func(result metrics.RequestResult) {
+				once.Do(func() {
+					metrics.TrackCompletedRequest(*r.URL, result)
+				})
+			}
+
+			contextWithCallback := context.WithValue(r.Context(), operations.CallbackFuncKey, callbackFunc)
+			r = r.WithContext(contextWithCallback)
+			metrics.TrackStartedRequest(*r.URL)
+			next.ServeHTTP(w, r)
+		})
+	})
+
+	mux.Use(func(next http.Handler) http.Handler {
```

Comment:
Regarding this, please correct me if I am wrong, but I believe we do not have direct access to the request context from .Render() unless we explicitely pass the request object to it.
If that is the case, I propose either:
1. Adding the request information to every relevant Response type, could be just the callback function instead of the entire request or request context. And then modify every relevant `Render` to extract the status code and call the callback function to track the request as completed.
2. Adding a Code() function to the Request interface to keep track of the resulting status code after Render() is called. Then the status code returned by Code() could be used to track the request using the callback function in the end of the handlers created by [createCmd](https://github.com/hamistao/lxd/blob/26163a66372da1905c0c3a3002321baba479e79c/lxd/daemon.go#L619). Using createCmd would cover all relevant endpoints for the API metrics, aside from the 404 not found and the '/' endpoint thqat would be handled separately.

Between those two, I think 2 looks better because it keeps the callback function call outside `Render` and delegates to the `Response` only the task of preserving the status code. What do you think?

Reply:
I think updating `Render` to take in the request as a parameter would also be an option.

Reply:
Yes you're right, I had only looked at the top two: `(r *eventsServe) Render(w http.ResponseWriter) error` and `(r *sftpServeResponse) Render(w http.ResponseWriter) error` which are special cases.

Adding a `Code` function (or a similar alternative of adding an int status code return value to Render itself) would not solve the issue of being able to detect the type of response (operation or synchronous) and mapping that code value to a metrics result type.

Also, the presence of `req` in `eventsServe` and `sftpServeResponse` (as well as some others) suggests that it may not be that much of a special case, and adding `r *http.Request` as an argument to `Render()` might allow us to standardise those outliers and bring them into the same pattern as the others.